ledger/complete/mtrie/forest.go

```diff
@@ -216,7 +216,8 @@ func (f *Forest) Proofs(r *ledger.TrieRead) (*ledger.TrieBatchProof, error) {
 
 	// if we have to insert empty values
 	if len(notFoundPaths) > 0 {
-		newTrie, err := trie.NewTrieWithUpdatedRegisters(stateTrie, notFoundPaths, notFoundPayloads)
+		// for proofs, we have to set the pruning to false, since wee need the trie to be expanded
```

Comment:
When we generate the proof for a register, there are generally two distinct cases: 
* the register is allocated: the proof for this register value is often called an "inclusion proof"
* the register is unallocated : the proof for this register value is often called an "non-inclusion proof"

Technically, a non-inclusion proof shows that, when descending the tree along the register path, you reach at some point a default hash. Then we know that the register is not allocated. Conceptually, this is quite straight forward. But from an engineering perspective, we haven't completely implemented the logic yet (a good amount of the logic is already there, but a few pieces are still missing ... mostly tests). 
What we do instead of delivering a properly formatted non-inclusion proof:
* We write the empty register payload into the trie. This is semantically equivalent to an unallocated register, i.e. the root hash remains invariant. But we force the trie to actually expand to this registers. Thereby, we replace all non-inclusion proofs by inclusion proof, where the included value is an empty register. 
* Mathematically, this is equivalent. But from an implementation perspective, we format the proof differently and also create significant computational overhead.

Reply:
the current implementation of batch proofs is based on inclusion proofs, so for values that don't exist, we need to expand the trie to be able to construct the inclusion proof. pruning would prune those branches and cause issues.